server/src/main/java/com/cloud/hypervisor/KVMGuru.java

```diff
@@ -68,6 +83,37 @@ protected KVMGuru() {
         super();
     }
 
+    /**
+     * Get next free DeviceId for a KVM Guest
+     */
+
+    protected Long getNextAvailableDeviceId(List<VolumeVO> vmVolumes) {
+
+        int maxDataVolumesSupported;
+        int maxDeviceId;
+        List<String> devIds = new ArrayList<>();
+
+        try {
+            maxDataVolumesSupported = _hypervisorCapabilitiesDao.getMaxDataVolumesLimit(HypervisorType.KVM,"default");
+            int maxDevices = maxDataVolumesSupported + 2; // add 2 to consider devices root volume and cdrom
```

Comment:
Can you help explain this logic, appears some hardcoded assumption?

Reply:
This is actually how /main/server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java@4070  handles the deviceid discovery. The +2 is KVM specific for anticipating root volume and removable cdrom devices as far as i understand.

Reply:
Thanks @fermosan makes sense.